src/job.rs

```diff
@@ -19,14 +105,59 @@ impl Job {
                     .collect::<Vec<&str>>()
                     .as_slice(),
             )
+            .current_dir(&work_dir)
+            // TODO: this is going to have to retain some environment variables
+            // for software to work correctly. For example, we'll probably need
+            // to provide a fake HOME the way Nix does.
+            .env_clear()
             .output()
+            .context("couldn't run the command")
+    }
+
+    fn tear_down_workspace(&self, work_dir: &Path) -> Result<()> {
+        for output in &self.outputs {
+            self.copy_creating_directories(
+                self.path_in_workspace(work_dir, &output)
+                    .context("couldn't determine path for output")?,
+                self.working_directory.join(output),
+            )
+            .context("couldn't retrieve output")?;
+        }
+
+        Ok(())
+    }
+
+    fn copy_creating_directories(&self, source: PathBuf, dest: PathBuf) -> Result<u64> {
+        match dest.parent() {
+                    Some(parent) => fs::create_dir_all(parent).with_context(|| format!("couldn't make the parent directories for {}", dest.display()))?,
+                    None => bail!("couldn't create the directories leading to {}. That probably means it's at the filesystem root, but we should have excluded that possibility already. This is a bug and should be reported.", dest.display())
+                };
+        fs::copy(&source, &dest)
+            .with_context(|| format!("couldn't copy {} to {}", source.display(), dest.display()))
+    }
+
+    fn path_in_workspace(&self, work_dir: &Path, input: &Path) -> Result<PathBuf> {
```

Comment:
📝 I was able to figure it out from reading it, but a brief comment about what this does would be nice since it's used in several places!

Reply:
documented in the struct fields now!